contract-derive/src/lib.rs

```diff
@@ -2,9 +2,89 @@ extern crate proc_macro;
 use alloy_core::primitives::keccak256;
 use proc_macro::TokenStream;
 use quote::{format_ident, quote};
-use syn::{parse_macro_input, ImplItem, ItemImpl};
+use syn::{parse_macro_input, ImplItem, ItemImpl, DeriveInput, Data, Fields};
 use syn::{FnArg, ReturnType};
 
+#[proc_macro_derive(Event, attributes(indexed))]
+pub fn event_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let name = &input.ident;
+    
+    let fields = if let Data::Struct(data) = &input.data {
+        if let Fields::Named(fields) = &data.fields {
+            &fields.named
+        } else {
+            panic!("Event must have named fields");
+        }
+    } else {
+        panic!("Event must be a struct");
+    };
+
+    // Collect iterators into vectors
+    let field_names: Vec<_> = fields.iter().map(|f| &f.ident).collect();
+    let field_types: Vec<_> = fields.iter().map(|f| &f.ty).collect();
+    let indexed_fields: Vec<_> = fields
+        .iter()
+        .filter(|f| {
+            f.attrs.iter().any(|attr| attr.path.is_ident("indexed"))
+        })
+        .map(|f| &f.ident)
+        .collect();
+
+    let expanded = quote! {
+        impl #name {
+            const NAME: &'static str = stringify!(#name);
+            const INDEXED_FIELDS: &'static [&'static str] = &[
+                #(stringify!(#indexed_fields)),*
+            ];
+
+            pub fn new(#(#field_names: #field_types),*) -> Self {
+                Self {
+                    #(#field_names),*
+                }
+            }
+        }
+
+        impl eth_riscv_runtime::log::Event for #name {
+            fn encode_log(&self) -> (alloc::vec::Vec<u8>, alloc::vec::Vec<[u8; 32]>) {
+                use alloy_sol_types::SolValue;
```

Comment:
I think there should be an even better way to do this here without manually creating the signature? Cc @DaniPopes